File: darts-suite/fortran/darts-omp-loop.f90
! Compute pi using OpenMP Loop
program darts
    use iso_fortran_env, only: real64
    use lcgenerator
    use omp_lib
    implicit none
    
    integer(kind=8) :: num_trials = 1000000_8
    integer(kind=8) :: ncirc = 0_8
    integer(kind=8) :: i, thread_seed, base_seed = 12345_8
    real(kind=8)    :: pi, x, y  ! Hoisted variables
    real(kind=8)    :: r = 1.0_8
    real(kind=8)    :: r2

    r2 = r * r

    !$omp parallel private(thread_seed, x, y) reduction(+:ncirc)
    !$omp loop
    do i = 1, num_trials
        thread_seed = base_seed + ((i-1) * 2_8)
        
        call lcgrandom_r(thread_seed, x)
        call lcgrandom_r(thread_seed, y)
        
        if ((x*x + y*y) <= r2) then
            ncirc = ncirc + 1_8
        end if
    end do
    !$omp end loop
    !$omp end parallel

    pi = 4.0_8 * real(ncirc, 8) / real(num_trials, 8)
    print*, '     '
    print*, '     Computing pi using OpenMP Loop:'
    print '(/,A,I0,A,F10.6,/)', '      For ', num_trials, ' trials, pi = ', pi
    print*, '     '
end program darts

File: darts-suite/c/darts-omp-loop.c
/* Compute pi using OpenMP "pragma omp loop" */ 

/*  Written by Helen He, June 2024
    In this code, we are not using the pseudorandom number generator used in other codes,
    instead we use a complete random number generator rand() in C
    so we do not need to use an OMP critial region (which is not allowed to nest under omp loop)
    to reproduce the randon seeds.  
    so the final pi value is not bit-for-bit from other codes
    Also, it gives a different pi value every time it runs
*/

#include "lcgenerator.h"
#include <omp.h>
#include <stdio.h>
#include <stdlib.h>

static long num_trials = 1000000; 

int main(int argc, char **argv) {
  long i;
  long Ncirc = 0;
  double pi, x, y;
  double r = 1.0; /* radius of circle */
  double r2 = r*r;
    
#pragma omp parallel private(x,y) reduction(+:Ncirc)
{
#pragma omp loop
  for (i = 0; i < num_trials; i++) {
     x = (double) rand() / RAND_MAX;
     y = (double) rand() / RAND_MAX;
     if ((x*x + y*y) <= r2)
        Ncirc++;
  }
}

  pi = 4.0 * ((double)Ncirc)/((double)num_trials);
  printf("\n \t Computing pi using OpenMP: \n");
  printf("\t For %ld trials, pi = %f\n", num_trials, pi);
  printf("\n");

  return 0;
}
